File: 두 큐 합 같게 만들기.js
function solution(queue1, queue2) {

    // 두 큐의 합을 같게 만들 수 없는 경우
    let answer = -1; 

    //reduce메소드를 이용해 큐의 원소를 모두 합함
    let sum1 = queue1.reduce((pre,cur)=>pre+cur,0); 
    let sum2 = queue2.reduce((pre,cur)=>pre+cur,0);

    // 두 큐를 모두 합한 값의 반
    const middle = (sum1+sum2)/2;

    // concat메소드를 이용한 큐1과 큐2를 합한 새로운 배열 temp 생성 
    let temp = queue1.concat(queue2);

    // 투포인터 left, right 이용
    let left = 0, right = queue1.length, count = 0;
    const len = temp.length;


    while(count < 2*len){
        if(sum1 > middle){
            sum1-=temp[left];
            if(left < len-1)
                left++;
            else
                left = 0;
        }
        else if(sum1<middle){
            sum1+=temp[right];
            if(right < len-1)
                right++;
            else
                right=0;
        }
        else
            return count;
        count++;
    }
    return answer;
}

File: 두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    answer = 0

    # deque 이용
    queue1, queue2 = deque(queue1), deque(queue2)
    sum1, sum2 = sum(queue1), sum(queue2)
    half = sum1 + sum2 // 2
    
    # 3번 돌아도 반값 못찾으면 return -1
    for _ in range(3*len(queue1)):

        # 왼쪽 배열의 원소의 합이 더 클 경우
        # 왼쪽 배열의 첫번째 원소를 오른쪽 배열의 맨 끝으로 이동
        if sum1 > sum2:
            sum1 -= queue1[0]
            sum2 += queue1[0]
            queue2.append(queue1.popleft())

        # 오른쪽 배열의 원소의 합이 더 클 경우
        # 오른쪽 배열의 첫번째 원소를 왼쪽 배열의 맨 끝으로 이동
        elif sum1 < sum2:
            sum1 += queue2[0]
            sum2 -= queue2[0]
            queue1.append(queue2.popleft())
        
        # 두 배열의 값이 같은 경우
        # 이동횟수를 담은 answer 반환
        else: 
            return answer
        answer += 1

    return -1